File: backend/quizzes.py
"""
Quiz routes - list, take, submit, grade quizzes
"""

from datetime import datetime
from flask import Blueprint, request, jsonify
from flask_login import current_user
from extensions import db
from backend.models import Quiz, QuizQuestion, QuizAttempt, QuizAnswer, StudentProgress
from backend.utils import require_login

quizzes_bp = Blueprint('quizzes', __name__, url_prefix='/api/quizzes')


@quizzes_bp.route('', methods=['GET'])
@require_login
def list_quizzes():
    """List available quizzes by grade and subject"""
    grade = request.args.get('grade', type=int)
    subject = request.args.get('subject', type=str)

    query = Quiz.query.filter_by(is_published=True)

    if grade:
        query = query.filter_by(grade_level=grade)
    if subject:
        query = query.filter_by(subject=subject)

    quizzes = query.all()

    return jsonify([{
        'id': q.id,
        'title': q.title,
        'description': q.description,
        'subject': q.subject,
        'grade_level': q.grade_level,
        'total_questions': q.total_questions,
        'passing_score': q.passing_score,
        'time_limit_minutes': q.time_limit_minutes,
        'created_at': q.created_at.isoformat()
    } for q in quizzes]), 200


@quizzes_bp.route('/<int:quiz_id>', methods=['GET'])
@require_login
def get_quiz(quiz_id):
    """Get quiz with questions (without correct answers)"""
    quiz = Quiz.query.get_or_404(quiz_id)

    if not quiz.is_published:
        return jsonify({'error': 'Quiz not available'}), 403

    questions = sorted(quiz.questions, key=lambda q: q.question_order)

    return jsonify({
        'id': quiz.id,
        'title': quiz.title,
        'description': quiz.description,
        'subject': quiz.subject,
        'grade_level': quiz.grade_level,
        'total_questions': quiz.total_questions,
        'time_limit_minutes': quiz.time_limit_minutes,
        'questions': [{
            'id': q.id,
            'question_text': q.question_text,
            'question_type': q.question_type,
            'option_a': q.option_a,
            'option_b': q.option_b,
            'option_c': q.option_c,
            'option_d': q.option_d,
            'marks': q.marks,
            'order': q.question_order
        } for q in questions]
    }), 200


@quizzes_bp.route('/<int:quiz_id>/start', methods=['POST'])
@require_login
def start_quiz(quiz_id):
    """Start a new quiz attempt"""
    if current_user.role != 'student':
        return jsonify({'error': 'Only students can take quizzes'}), 403

    quiz = Quiz.query.get_or_404(quiz_id)

    # Check for in-progress attempt
    existing = QuizAttempt.query.filter_by(
        student_id=current_user.id,
        quiz_id=quiz_id,
        status='in_progress'
    ).first()

    if existing:
        return jsonify({
            'attempt_id': existing.id,
            'started_at': existing.started_at.isoformat(),
            'message': 'Resuming existing attempt'
        }), 200

    # Create new attempt
    attempt = QuizAttempt(
        student_id=current_user.id,
        quiz_id=quiz_id,
        status='in_progress'
    )
    db.session.add(attempt)
    db.session.commit()

    return jsonify({
        'attempt_id': attempt.id,
        'started_at': attempt.started_at.isoformat(),
        'time_limit_minutes': quiz.time_limit_minutes
    }), 201


@quizzes_bp.route('/<int:quiz_id>/submit', methods=['POST'])
@require_login
def submit_quiz(quiz_id):
    """Submit quiz answers and get graded"""
    if current_user.role != 'student':
        return jsonify({'error': 'Only students can submit quizzes'}), 403

    data = request.get_json()
    if not data or 'answers' not in data:
        return jsonify({'error': 'Missing answers'}), 400

    quiz = Quiz.query.get_or_404(quiz_id)
    attempt_id = data.get('attempt_id')

    # Get or create attempt
    if attempt_id:
        attempt = QuizAttempt.query.get(attempt_id)
        if not attempt or attempt.student_id != current_user.id:
            return jsonify({'error': 'Invalid attempt'}), 403
    else:
        attempt = QuizAttempt(
            student_id=current_user.id,
            quiz_id=quiz_id,
            status='in_progress'
        )
        db.session.add(attempt)
        db.session.flush()

    # Grade answers
    answers = data['answers']
    total_score = 0.0
    total_marks = 0.0
    questions_dict = {q.id: q for q in quiz.questions}
    needs_review = False

    for ans in answers:
        question_id = ans.get('question_id')
        student_answer = ans.get('answer', '').strip()

        if question_id not in questions_dict:
            continue

        question = questions_dict[question_id]
        total_marks += question.marks

        # Auto-grade MCQ
        if question.question_type == 'mcq':
            is_correct = (student_answer.upper() == question.correct_option)
            marks_awarded = question.marks if is_correct else 0.0
            total_score += marks_awarded

            answer_record = QuizAnswer(
                attempt_id=attempt.id,
                question_id=question_id,
                student_answer=student_answer,
                is_correct=is_correct,
                marks_awarded=marks_awarded,
                graded_at=datetime.utcnow()
            )
        # Auto-grade short answer (simple exact match)
        elif question.question_type == 'short_answer':
            expected = (question.expected_answer or '').strip().lower()
            given = student_answer.lower()
            is_correct = (expected == given) if expected else None
            marks_awarded = question.marks if is_correct else 0.0
            total_score += marks_awarded

            answer_record = QuizAnswer(
                attempt_id=attempt.id,
                question_id=question_id,
                student_answer=student_answer,
                is_correct=is_correct,
                marks_awarded=marks_awarded,
                graded_at=datetime.utcnow()
            )
        # Essay - needs teacher review
        else:
            needs_review = True
            answer_record = QuizAnswer(
                attempt_id=attempt.id,
                question_id=question_id,
                student_answer=student_answer,
                is_correct=None,
                marks_awarded=0.0
            )

        db.session.add(answer_record)

    # Update attempt
    attempt.score = total_score
    attempt.total_marks = total_marks
    attempt.percentage = (total_score / total_marks * 100) if total_marks > 0 else 0.0
    attempt.submitted_at = datetime.utcnow()
    attempt.status = 'graded' if not needs_review else 'submitted'

    # Update student progress
    update_student_progress(
        current_user.id,
        quiz.subject,
        quiz.grade_level,
        correct_count=sum(1 for a in answers if a.get('question_id') in questions_dict
                          and questions_dict[a['question_id']].question_type == 'mcq'
                          and a.get('answer', '').upper() == questions_dict[a['question_id']].correct_option),
        total_count=len([a for a in answers if a.get('question_id') in questions_dict
                         and questions_dict[a['question_id']].question_type == 'mcq']),
        quiz_score=attempt.percentage
    )

    db.session.commit()

    return jsonify({
        'attempt_id': attempt.id,
        'score': total_score,
        'total_marks': total_marks,
        'percentage': attempt.percentage,
        'status': attempt.status,
        'passed': attempt.percentage >= quiz.passing_score,
        'needs_review': needs_review
    }), 200


@quizzes_bp.route('/<int:quiz_id>/results/<int:attempt_id>', methods=['GET'])
@require_login
def get_results(quiz_id, attempt_id):
    """Get detailed quiz results"""
    attempt = QuizAttempt.query.get_or_404(attempt_id)

    if attempt.student_id != current_user.id and current_user.role != 'teacher':
        return jsonify({'error': 'Access denied'}), 403

    quiz = Quiz.query.get_or_404(quiz_id)
    answer_details = []

    for ans in attempt.answers:
        q = ans.question
        answer_details.append({
            'question_id': q.id,
            'question_text': q.question_text,
            'question_type': q.question_type,
            'student_answer': ans.student_answer,
            'correct_answer': q.correct_option if q.question_type == 'mcq' else q.expected_answer,
            'is_correct': ans.is_correct,
            'marks_awarded': ans.marks_awarded,
            'max_marks': q.marks,
            'explanation': q.explanation
        })

    return jsonify({
        'attempt_id': attempt.id,
        'quiz_title': quiz.title,
        'score': attempt.score,
        'total_marks': attempt.total_marks,
        'percentage': attempt.percentage,
        'status': attempt.status,
        'submitted_at': attempt.submitted_at.isoformat() if attempt.submitted_at else None,
        'answers': answer_details
    }), 200


@quizzes_bp.route('/create', methods=['POST'])
@require_login
def create_quiz():
    """Create a new quiz (teacher only)"""
    if current_user.role != 'teacher':
        return jsonify({'error': 'Only teachers can create quizzes'}), 403

    data = request.get_json()
    if not data:
        return jsonify({'error': 'Missing data'}), 400

    required = ['title', 'subject', 'grade_level', 'questions']
    for field in required:
        if field not in data:
            return jsonify({'error': f'Missing required field: {field}'}), 400

    if not isinstance(data['questions'], list) or len(data['questions']) == 0:
        return jsonify({'error': 'Questions must be a non-empty list'}), 400

    try:
        quiz = Quiz(
            title=data['title'],
            description=data.get('description', ''),
            subject=data['subject'],
            grade_level=data['grade_level'],
            created_by=current_user.id,
            total_questions=len(data['questions']),
            passing_score=data.get('passing_score', 60.0),
            time_limit_minutes=data.get('time_limit_minutes'),
            is_published=True
        )
        db.session.add(quiz)
        db.session.flush()

        for idx, q_data in enumerate(data['questions']):
            question = QuizQuestion(
                quiz_id=quiz.id,
                question_text=q_data['question_text'],
                question_type=q_data.get('question_type', 'mcq'),
                option_a=q_data.get('option_a'),
                option_b=q_data.get('option_b'),
                option_c=q_data.get('option_c'),
                option_d=q_data.get('option_d'),
                correct_option=q_data.get('correct_option'),
                expected_answer=q_data.get('expected_answer'),
                explanation=q_data.get('explanation'),
                question_order=idx,
                marks=q_data.get('marks', 1.0)
            )
            db.session.add(question)

        db.session.commit()
        return jsonify({
            'message': 'Quiz created successfully',
            'quiz_id': quiz.id,
            'title': quiz.title
        }), 201

    except Exception as e:
        db.session.rollback()
        return jsonify({'error': f'Failed to create quiz: {str(e)}'}), 500


@quizzes_bp.route('/my-attempts', methods=['GET'])
@require_login
def my_attempts():
    """Get student's quiz attempts"""
    attempts = QuizAttempt.query.filter_by(student_id=current_user.id).all()

    return jsonify([{
        'attempt_id': a.id,
        'quiz_id': a.quiz_id,
        'quiz_title': a.quiz.title,
        'subject': a.quiz.subject,
        'score': a.score,
        'percentage': a.percentage,
        'status': a.status,
        'submitted_at': a.submitted_at.isoformat() if a.submitted_at else None
    } for a in attempts]), 200


def update_student_progress(student_id, subject, grade_level, correct_count, total_count, quiz_score):
    """Update or create student progress record"""
    progress = StudentProgress.query.filter_by(
        student_id=student_id,
        subject=subject
    ).first()

    if not progress:
        progress = StudentProgress(
            student_id=student_id,
            subject=subject,
            grade_level=grade_level,
            total_questions_answered=0,
            correct_answers=0,
            quizzes_completed=0,
            total_score=0.0,
            doubts_resolved=0,
            resources_viewed=0
        )
        db.session.add(progress)
        db.session.flush()

    progress.total_questions_answered = (progress.total_questions_answered or 0) + total_count
    progress.correct_answers = (progress.correct_answers or 0) + correct_count
    progress.quizzes_completed = (progress.quizzes_completed or 0) + 1
    progress.total_score = (progress.total_score or 0.0) + (quiz_score or 0.0)
    progress.last_activity = datetime.utcnow()

File: frontend/js/quizzes.js
/**
 * Quiz panel handlers - taking quizzes, viewing results, creating quizzes
 */

let currentQuizzes = [];
let currentQuiz = null;
let currentAttempt = null;
let quizAnswers = {};
let questionCounter = 0;

/**
 * Load and display quizzes
 */
async function loadQuizzes() {
    const user = Utils.getCurrentUser();
    if (!user) return;

    try {
        const subject = document.getElementById('quizSubjectFilter')?.value || null;
        // Use grade selector for students
        const grade = typeof getCurrentViewingGrade === 'function' ? getCurrentViewingGrade() : (user.role === 'student' ? user.gradeLevel : null);

        const quizzes = await API.quizzes.list(grade, subject);

        if (quizzes.length) {
            await offlineDB.cache('/quizzes', quizzes);
        }

        currentQuizzes = quizzes;
        displayQuizzes(quizzes);
    } catch (error) {
        console.error('Failed to load quizzes:', error);
        const cached = await offlineDB.getCached('/quizzes');
        if (cached) displayQuizzes(cached);
    }
}

/**
 * Display quizzes
 */
function displayQuizzes(quizzes) {
    const container = document.getElementById('quizList');

    if (!quizzes || quizzes.length === 0) {
        container.innerHTML = '<p>No quizzes available. ' +
            (Utils.isTeacher() ? 'Click "Create Quiz" to add one!' : 'Check back later!') + '</p>';
        return;
    }

    container.innerHTML = quizzes.map(quiz => `
        <div class="resource-card">
            <h3>${quiz.title}</h3>
            <p>${quiz.description || 'No description'}</p>
            <div class="resource-meta">
                <span>${quiz.subject}</span>
                <span>Grade ${quiz.grade_level}</span>
                <span>${quiz.total_questions} questions</span>
            </div>
            ${quiz.time_limit_minutes ? `<p><strong>Time Limit:</strong> ${quiz.time_limit_minutes} min</p>` : ''}
            <div class="resource-actions">
                ${Utils.isStudent()
                    ? `<button onclick="startQuiz(${quiz.id})">Start Quiz</button>`
                    : `<button onclick="viewQuizDetails(${quiz.id})">View Details</button>`
                }
            </div>
        </div>
    `).join('');
}

/**
 * Start taking a quiz
 */
async function startQuiz(quizId) {
    try {
        // Get quiz details
        currentQuiz = await API.quizzes.get(quizId);

        // Start attempt
        const attempt = await API.quizzes.startAttempt(quizId);
        currentAttempt = attempt;
        quizAnswers = {};

        // Show quiz taking view
        document.getElementById('quizListView').classList.add('hidden');
        document.getElementById('quizTakingView').classList.remove('hidden');
        document.getElementById('quizCreateView').classList.add('hidden');
        document.getElementById('quizResultsView').classList.add('hidden');

        document.getElementById('quizTitle').textContent = currentQuiz.title;
        document.getElementById('quizProgress').textContent =
            `0/${currentQuiz.questions.length} answered`;

        renderQuestions(currentQuiz.questions);
    } catch (error) {
        console.error('Failed to start quiz:', error);
        Utils.showSuccess('Failed to start quiz: ' + (error.message || 'Unknown error'));
    }
}

/**
 * Render quiz questions
 */
function renderQuestions(questions) {
    const container = document.getElementById('quizQuestions');
    container.innerHTML = questions.map((q, idx) => `
        <div class="quiz-question" data-question-id="${q.id}">
            <h4>Q${idx + 1}. ${q.question_text}</h4>
            ${q.question_type === 'mcq' ? `
                <div class="quiz-options">
                    ${q.option_a ? `
                        <label class="quiz-option">
                            <input type="radio" name="q${q.id}" value="A" onchange="updateAnswer(${q.id}, 'A')">
                            <span><strong>A.</strong> ${q.option_a}</span>
                        </label>
                    ` : ''}
                    ${q.option_b ? `
                        <label class="quiz-option">
                            <input type="radio" name="q${q.id}" value="B" onchange="updateAnswer(${q.id}, 'B')">
                            <span><strong>B.</strong> ${q.option_b}</span>
                        </label>
                    ` : ''}
                    ${q.option_c ? `
                        <label class="quiz-option">
                            <input type="radio" name="q${q.id}" value="C" onchange="updateAnswer(${q.id}, 'C')">
                            <span><strong>C.</strong> ${q.option_c}</span>
                        </label>
                    ` : ''}
                    ${q.option_d ? `
                        <label class="quiz-option">
                            <input type="radio" name="q${q.id}" value="D" onchange="updateAnswer(${q.id}, 'D')">
                            <span><strong>D.</strong> ${q.option_d}</span>
                        </label>
                    ` : ''}
                </div>
            ` : q.question_type === 'short_answer' ? `
                <input type="text" placeholder="Your answer..."
                    onchange="updateAnswer(${q.id}, this.value)"
                    class="quiz-text-input">
            ` : `
                <textarea rows="4" placeholder="Write your answer..."
                    onchange="updateAnswer(${q.id}, this.value)"
                    class="quiz-text-input"></textarea>
            `}
        </div>
    `).join('');
}

/**
 * Update answer for a question
 */
function updateAnswer(questionId, value) {
    quizAnswers[questionId] = value;
    const answered = Object.keys(quizAnswers).length;
    const total = currentQuiz.questions.length;
    document.getElementById('quizProgress').textContent = `${answered}/${total} answered`;
}

/**
 * Submit quiz
 */
async function submitQuiz() {
    if (!currentQuiz || !currentAttempt) return;

    const answersList = Object.entries(quizAnswers).map(([qid, ans]) => ({
        question_id: parseInt(qid),
        answer: ans
    }));

    if (answersList.length === 0) {
        if (!confirm('You have not answered any questions. Submit anyway?')) {
            return;
        }
    }

    try {
        const result = await API.quizzes.submit(currentQuiz.id, currentAttempt.attempt_id, answersList);
        displayQuizResults(result);
    } catch (error) {
        console.error('Failed to submit quiz:', error);
        // Try to save offline
        if (!navigator.onLine) {
            await offlineDB.saveQuizAnswers(currentQuiz.id, {
                attempt_id: currentAttempt.attempt_id,
                answers: answersList
            });
            Utils.showSuccess('Quiz saved offline. Will be submitted when online.');
        } else {
            Utils.showSuccess('Failed to submit: ' + (error.message || 'Unknown error'));
        }
    }
}

/**
 * Display quiz results
 */
async function displayQuizResults(result) {
    document.getElementById('quizTakingView').classList.add('hidden');
    document.getElementById('quizResultsView').classList.remove('hidden');

    // Get detailed results
    let details = null;
    try {
        details = await API.quizzes.results(currentQuiz.id, result.attempt_id);
    } catch (e) {
        console.error('Failed to load detailed results:', e);
    }

    const container = document.getElementById('quizResults');
    const passClass = result.passed ? 'pass' : 'fail';

    let html = `
        <div class="quiz-result-summary ${passClass}">
            <h2>${result.passed ? 'Well Done!' : 'Keep Practicing!'}</h2>
            <div class="score-display">
                <span class="score-percentage">${result.percentage.toFixed(1)}%</span>
                <span class="score-fraction">${result.score} / ${result.total_marks}</span>
            </div>
            <p>${result.passed ? 'You passed!' : 'You can do better next time.'}</p>
            ${result.needs_review ? '<p>Some answers need teacher review.</p>' : ''}
        </div>
    `;

    if (details && details.answers) {
        html += '<h3>Detailed Results</h3>';
        details.answers.forEach((ans, idx) => {
            const correctClass = ans.is_correct === true ? 'correct' :
                                 ans.is_correct === false ? 'incorrect' : 'pending';
            html += `
                <div class="answer-review ${correctClass}">
                    <h4>Q${idx + 1}. ${ans.question_text}</h4>
                    <p><strong>Your answer:</strong> ${ans.student_answer || '(no answer)'}</p>
                    ${ans.question_type === 'mcq' && ans.correct_answer ?
                        `<p><strong>Correct answer:</strong> ${ans.correct_answer}</p>` : ''}
                    <p><strong>Marks:</strong> ${ans.marks_awarded} / ${ans.max_marks}</p>
                    ${ans.explanation ? `<p class="explanation"><strong>Explanation:</strong> ${ans.explanation}</p>` : ''}
                </div>
            `;
        });
    }

    container.innerHTML = html;
}

/**
 * Back to quiz list
 */
function backToQuizList() {
    document.getElementById('quizListView').classList.remove('hidden');
    document.getElementById('quizTakingView').classList.add('hidden');
    document.getElementById('quizResultsView').classList.add('hidden');
    document.getElementById('quizCreateView').classList.add('hidden');
    currentQuiz = null;
    currentAttempt = null;
    quizAnswers = {};
    loadQuizzes();
}

/**
 * Show quiz creation form (teacher only)
 */
function showCreateQuizForm() {
    document.getElementById('quizListView').classList.add('hidden');
    document.getElementById('quizCreateView').classList.remove('hidden');
    questionCounter = 0;
    document.getElementById('questionsContainer').innerHTML = '';
    addQuestion();
}

/**
 * Add a question to the create quiz form
 */
function addQuestion() {
    questionCounter++;
    const container = document.getElementById('questionsContainer');
    const qDiv = document.createElement('div');
    qDiv.className = 'create-question';
    qDiv.dataset.qid = questionCounter;
    qDiv.innerHTML = `
        <h4>Question ${questionCounter}</h4>
        <textarea placeholder="Question text" class="q-text" rows="2" required></textarea>
        <select class="q-type" onchange="toggleOptions(this)">
            <option value="mcq">Multiple Choice</option>
            <option value="short_answer">Short Answer</option>
            <option value="essay">Essay</option>
        </select>
        <div class="mcq-options">
            <input type="text" placeholder="Option A" class="q-opt-a">
            <input type="text" placeholder="Option B" class="q-opt-b">
            <input type="text" placeholder="Option C" class="q-opt-c">
            <input type="text" placeholder="Option D" class="q-opt-d">
            <select class="q-correct">
                <option value="">Correct Option</option>
                <option value="A">A</option>
                <option value="B">B</option>
                <option value="C">C</option>
                <option value="D">D</option>
            </select>
        </div>
        <input type="text" placeholder="Expected answer (for short answer)" class="q-expected hidden">
        <textarea placeholder="Explanation (optional)" class="q-explanation" rows="2"></textarea>
        <button type="button" onclick="removeQuestion(${questionCounter})" class="btn btn-danger btn-small">Remove</button>
        <hr>
    `;
    container.appendChild(qDiv);
}

/**
 * Toggle options visibility based on question type
 */
function toggleOptions(select) {
    const parent = select.closest('.create-question');
    const mcqOpts = parent.querySelector('.mcq-options');
    const expected = parent.querySelector('.q-expected');

    if (select.value === 'mcq') {
        mcqOpts.classList.remove('hidden');
        expected.classList.add('hidden');
    } else if (select.value === 'short_answer') {
        mcqOpts.classList.add('hidden');
        expected.classList.remove('hidden');
    } else {
        mcqOpts.classList.add('hidden');
        expected.classList.add('hidden');
    }
}

/**
 * Remove a question
 */
function removeQuestion(qid) {
    const el = document.querySelector(`[data-qid="${qid}"]`);
    if (el) el.remove();
}

/**
 * Create the quiz
 */
async function createQuiz(e) {
    e.preventDefault();

    const title = document.getElementById('quizTitleInput').value.trim();
    const description = document.getElementById('quizDescriptionInput').value.trim();
    const subject = document.getElementById('quizSubjectInput').value;
    const grade = parseInt(document.getElementById('quizGradeInput').value);
    const timeLimit = document.getElementById('quizTimeLimitInput').value;

    // Collect questions
    const questionEls = document.querySelectorAll('.create-question');
    const questions = [];

    for (const qEl of questionEls) {
        const qText = qEl.querySelector('.q-text').value.trim();
        const qType = qEl.querySelector('.q-type').value;

        if (!qText) continue;

        const q = {
            question_text: qText,
            question_type: qType,
            explanation: qEl.querySelector('.q-explanation').value.trim()
        };

        if (qType === 'mcq') {
            q.option_a = qEl.querySelector('.q-opt-a').value.trim();
            q.option_b = qEl.querySelector('.q-opt-b').value.trim();
            q.option_c = qEl.querySelector('.q-opt-c').value.trim();
            q.option_d = qEl.querySelector('.q-opt-d').value.trim();
            q.correct_option = qEl.querySelector('.q-correct').value;

            if (!q.correct_option) {
                Utils.showSuccess('Please select correct option for all MCQ questions');
                return;
            }
        } else if (qType === 'short_answer') {
            q.expected_answer = qEl.querySelector('.q-expected').value.trim();
        }

        questions.push(q);
    }

    if (questions.length === 0) {
        Utils.showSuccess('Please add at least one question');
        return;
    }

    try {
        const result = await API.quizzes.create({
            title, description, subject,
            grade_level: grade,
            time_limit_minutes: timeLimit ? parseInt(timeLimit) : null,
            questions
        });

        Utils.showSuccess('Quiz created successfully!');
        document.getElementById('createQuizForm').reset();
        backToQuizList();
    } catch (error) {
        console.error('Failed to create quiz:', error);
        Utils.showSuccess('Failed to create quiz: ' + (error.message || 'Unknown error'));
    }
}

/**
 * Setup event listeners
 */
document.addEventListener('DOMContentLoaded', () => {
    const quizzesBtn = document.getElementById('quizzesBtn');
    if (quizzesBtn) {
        quizzesBtn.addEventListener('click', () => {
            switchPanel('quizzesPanel', quizzesBtn);
            backToQuizList();

            // Show create button for teachers
            if (Utils.isTeacher()) {
                document.getElementById('createQuizBtn').classList.remove('hidden');
            }
        });
    }

    document.getElementById('quizSubjectFilter')?.addEventListener('change', loadQuizzes);
    document.getElementById('refreshQuizzes')?.addEventListener('click', loadQuizzes);
    document.getElementById('createQuizBtn')?.addEventListener('click', showCreateQuizForm);
    document.getElementById('submitQuizBtn')?.addEventListener('click', submitQuiz);
    document.getElementById('cancelQuizBtn')?.addEventListener('click', backToQuizList);
    document.getElementById('backToQuizzesBtn')?.addEventListener('click', backToQuizList);
    document.getElementById('cancelCreateQuizBtn')?.addEventListener('click', backToQuizList);
    document.getElementById('addQuestionBtn')?.addEventListener('click', addQuestion);
    document.getElementById('createQuizForm')?.addEventListener('submit', createQuiz);
});
